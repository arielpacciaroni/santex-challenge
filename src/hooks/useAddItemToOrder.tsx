import { gql, useApolloClient, useMutation } from '@apollo/client';
import { addItemToOrderMutation } from '../graphql/mutations';
import { ProductVariant } from '../interfaces/ProductVariant';
import { useContext } from 'react';
import ProductContext from '../components/ProductContext';

interface ItemParams {
  productVariantId: ProductVariant['id'];
  quantity: number;
}

export function useAddItemToOrder() {
  const context = useContext(ProductContext);
  const queryClient = useApolloClient();

  const [mutateAddItemToOrder] = useMutation(addItemToOrderMutation);

  const addItemToOrder = async (params: ItemParams) => {
    await mutateAddItemToOrder({ variables: params });
    const variant = queryClient.readFragment<ProductVariant>({
      id: `ProductVariant:${params.productVariantId}`,
      fragment: gql`
        fragment Variant on ProductVariant {
          id
          name
          price
        }
      `,
    });

    if (variant) {
      const subTotal = context.subTotal;
      context.setSubTotal(subTotal + variant.price * params.quantity);
    }
  };

  return {
    addItemToOrder,
  };
}
