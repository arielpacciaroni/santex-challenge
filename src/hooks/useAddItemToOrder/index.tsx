import { useApolloClient, useMutation } from '@apollo/client';
import { addItemToOrderMutation } from '../../graphql/mutations';
import { ProductVariant } from '../../interfaces/ProductVariant';
import { useContext } from 'react';
import ProductContext from '../../components/ProductContext';
import { productVariantFragment } from '../../graphql/queries';

interface ItemParams {
  productVariantId: ProductVariant['id'];
  quantity: number;
}

export function useAddItemToOrder() {
  const context = useContext(ProductContext);
  const queryClient = useApolloClient();

  const [mutateAddItemToOrder] = useMutation(addItemToOrderMutation);

  const addItemToOrder = async (params: ItemParams) => {
    await mutateAddItemToOrder({ variables: params });
    const variant = queryClient.readFragment<ProductVariant>({
      id: `ProductVariant:${params.productVariantId}`,
      fragment: productVariantFragment,
    });

    if (variant) {
      const subTotal = context.subTotal;
      context.setSubTotal(subTotal + variant.price * params.quantity);
    }
  };

  return {
    addItemToOrder,
  };
}
