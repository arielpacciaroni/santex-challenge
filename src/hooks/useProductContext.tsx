import { useContext } from "react";
import ProductContext from "../components/ProductContext";

export function useProductContext() {
  return useContext(ProductContext);
}