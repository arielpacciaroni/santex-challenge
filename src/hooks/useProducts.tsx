import { useQuery } from '@apollo/client';
import { Product, productsQuery } from '../graphql/queries';
import { useMemo } from 'react';

interface ProductParams {
  page: number;
  itemsPerPage: number;
}

interface ProductResults {
  products: {
    items: Product[];
    totalItems: number;
  };
}

export function useProducts(params: ProductParams) {
  const { page, itemsPerPage } = params;

  const { data, loading, error, fetchMore } = useQuery<ProductResults>(
    productsQuery,
    {
      variables: {
        limit: itemsPerPage * page,
        offset: 0,
      },
    }
  );

  const hasNextPage = useMemo(() => {
    if (!data) return false;

    return data.products.totalItems > itemsPerPage * page;
  }, [data, page, itemsPerPage]);

  return {
    data: data?.products,
    loading,
    error,
    fetchMore,
    hasNextPage,
  };
}
