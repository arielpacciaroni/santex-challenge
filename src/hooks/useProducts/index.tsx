import { useLazyQuery } from '@apollo/client';
import { Product } from '../../interfaces/Product';
import { productsQuery } from '../../graphql/queries';

interface ProductResults {
  products: {
    items: Product[];
    totalItems: number;
  };
}

export function useProducts() {
  const [fetchMore, { called, loading, data, error }] =
    useLazyQuery<ProductResults>(productsQuery, {
      variables: {
        limit: 6,
        offset: 0,
      },
    });

  return {
    data: data?.products,
    totalItems: data?.products.totalItems ?? 0,
    loading,
    error,
    called,
    fetchMore,
  };
}
