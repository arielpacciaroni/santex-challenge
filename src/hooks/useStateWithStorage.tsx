import { useState } from "react";

export default function useStateWithStorage<T>(
  key: string,
  defaultValue: T
) {
  const [value, setValue] = useState<T>(() => {
    const storedValue = localStorage.getItem(key);
    return storedValue !== null ? JSON.parse(storedValue) : defaultValue;
  });

  const setValueWithStorage = (newValue: T) => {
    setValue(newValue);
    localStorage.setItem(key, JSON.stringify(newValue));
  };

  return [value, setValueWithStorage] as const;
}

