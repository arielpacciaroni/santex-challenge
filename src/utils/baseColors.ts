export const baseColors = {
  "light-gray": "#f5f5f5",
  "highlight": "#eee",
  "success": "#15803d",
  "secondary": '#1c1917',
};