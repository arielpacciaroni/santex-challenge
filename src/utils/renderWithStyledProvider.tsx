import { render } from "@testing-library/react";
import { ReactNode } from "react"
import { ThemeProvider } from "styled-components";
import { theme } from "./theme";

export function renderWithStyledProvider(element: ReactNode) {
  return render(
    <ThemeProvider theme={theme}>
      {element}
    </ThemeProvider>
  )
}