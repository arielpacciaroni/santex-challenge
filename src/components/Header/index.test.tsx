import { Header } from ".";
import { useProductContext } from "../../hooks/useProductContext";
import { renderWithStyledProvider } from "../../utils/renderWithStyledProvider";
import { screen } from "@testing-library/react";

jest.mock('../../hooks/useProductContext');

describe('Header component', () => {
  it('should render the header', () => {
    (useProductContext as jest.MockedFunction<typeof useProductContext>).mockReturnValue({
      subTotal: 0,
      setSubTotal: jest.fn()
    });

    const component = renderWithStyledProvider(
      <Header />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render the correct subTotal in the header', () => {
    (useProductContext as jest.MockedFunction<typeof useProductContext>).mockReturnValue({
      subTotal: 100,
      setSubTotal: jest.fn()
    });

    renderWithStyledProvider(
      <Header />
    );

    expect(screen.queryByText('$100')).toBeInTheDocument();
  })
})