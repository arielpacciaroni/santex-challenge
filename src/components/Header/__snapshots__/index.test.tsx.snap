// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render the header 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c1 {
  width: 100%;
  max-width: 720px;
  margin: 0 auto;
}

.c4 {
  border: none;
  outline: none;
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  font-size: 0.875rem;
  font-weight: 600;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  -webkit-transition-delay: 100ms;
  transition-delay: 100ms;
  cursor: pointer;
  color: #fff;
  background-color: #9a3412;
}

.c4:hover {
  background-color: #ea580c;
}

.c4:disabled {
  background-color: #eee;
  color: #aaa;
  cursor: not-allowed;
}

.c0 {
  background-color: #171717;
  color: #eee;
  padding: 1rem;
}

.c3 {
  width: 100px;
  height: auto;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (min-width:520px) {
  .c2 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

<body>
    <div>
      <header
        class="c0"
      >
        <div
          class="c1"
        >
          <section
            class="c2"
          >
            <div>
              <img
                alt="Logo"
                class="c3"
                src="https://santex.wpengine.com/wp-content/uploads/2019/02/logo-santex@3x.png"
              />
              <p>
                Your subtotal: $
                0
              </p>
            </div>
            <button
              class="c4"
            >
              Clear cart
            </button>
          </section>
        </div>
      </header>
    </div>
  </body>,
  "container": .c1 {
  width: 100%;
  max-width: 720px;
  margin: 0 auto;
}

.c4 {
  border: none;
  outline: none;
  padding: 0.5rem 1rem;
  border-radius: 0.5rem;
  font-size: 0.875rem;
  font-weight: 600;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  -webkit-transition-delay: 100ms;
  transition-delay: 100ms;
  cursor: pointer;
  color: #fff;
  background-color: #9a3412;
}

.c4:hover {
  background-color: #ea580c;
}

.c4:disabled {
  background-color: #eee;
  color: #aaa;
  cursor: not-allowed;
}

.c0 {
  background-color: #171717;
  color: #eee;
  padding: 1rem;
}

.c3 {
  width: 100px;
  height: auto;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media (min-width:520px) {
  .c2 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

<div>
    <header
      class="c0"
    >
      <div
        class="c1"
      >
        <section
          class="c2"
        >
          <div>
            <img
              alt="Logo"
              class="c3"
              src="https://santex.wpengine.com/wp-content/uploads/2019/02/logo-santex@3x.png"
            />
            <p>
              Your subtotal: $
              0
            </p>
          </div>
          <button
            class="c4"
          >
            Clear cart
          </button>
        </section>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
